### PR TITLE
(WIP) MAID-3164: Make interesting event handling fork resilient

### PR DESCRIPTION
**Pull request by madadam**
**DO NOT MERGE**. PR paused until https://maidsafe.atlassian.net/browse/MAID-3203 is implemented.

Notes for reviewers:

This PR deviates somehow from the description in https://maidsafe.atlassian.net/browse/MAID-3164, but I believe it should be functionally equivalent. In particular:

- Instead of the proposed `interesting` field in `Event`, there is only `consensus: Option<usize>` which is used to mark consensused interesting events. It is equivalent to the `Consensused(usize)` variant. The remaining states are inferred from whether `interesting_content` is empty (`NotInteresting`) or not (`Interesting`).
- `Parsec::interesting_events` is still around and used like before.
- `Parsec::next_stable_block` is modified to filter out interesting events that the deciding event does no longer see and those with empty `interesting_content`. This should achieve the same goal as the task requires.

The reason for this deviation is that it is much less invasive, in particular, it requires no changes to the dot parser and the dot files.

Unless I'm completely wrong :grimacing:

### CODE REVIEW

File: src/parsec.rs [hunk 1]
```diff
@@ -488,6 +490,12 @@ impl<T: NetworkEvent, S: SecretId> Parsec<T, S> {
             let observation = block.payload().clone();
             self.consensused_blocks.push_back(block);
 
+            // Mark the interesting event as consensused.
+            let next_stable_block_index = self.consensus_history.len();
+            if let Ok(event) = self.get_known_event_mut(&interesting_event_hash) {
+                event.consensus = Some(next_stable_block_index);
+            }
```
Comment (fizyk20) [inline src/parsec.rs]: I think we should be setting the `consensus` field in multiple events - all the interesting ones that carried the consensused payload, or even the ones that contained the payload (the ones that the interesting events see), I'm not even sure at the moment...

File: src/parsec.rs [hunk 2]
```diff
@@ -951,66 +959,120 @@ impl<T: NetworkEvent, S: SecretId> Parsec<T, S> {
...
+
+        Some((winning_hash, block))
+    }
+
+    // Returns an iterator over the decided meta-votes for the given observer
+    // event. Returns `None` if there is not enough meta-votes yet.
+    fn decided_meta_votes(
+        &self,
+        event_hash: &Hash,
+    ) -> Option<impl Iterator<Item = (&S::PublicId, bool)>> {
```
Comment (fizyk20) [inline src/parsec.rs]: Why not `Option<impl Iterator<Item = &S::PublicId>>`, and only iterate through peers that got `true`?

File: src/parsec.rs [hunk 3]
```diff
@@ -951,66 +959,120 @@ impl<T: NetworkEvent, S: SecretId> Parsec<T, S> {
...
+    {
+        let deciding_event = if let Ok(event) = self.get_known_event(event_hash) {
+            event
+        } else {
+            return vec![];
+        };
+
+        meta_votes
+            .into_iter()
+            .filter_map(|(id, decision)| {
```
Comment (fizyk20) [inline src/parsec.rs]: See the other comment - we omit the peers that have `false`, anyway.

File: src/parsec.rs (hunk 2 above)
Comment (madadam) [reply src/parsec.rs]: :+1: Good catch!

File: src/parsec.rs (hunk 3 above)
Comment (madadam) [reply src/parsec.rs]: :+1:

**Comment by madadam** [comment]
Closing for being stale. Will be superseded by a different PR in the future.

**Pull request closed without merging**